modules/oit/glsl/oit/volumeresolve.frag

```diff
@@ -0,0 +1,339 @@
+            }
+
+            // determine depth of next fragment to define the current ray segment for raycasting
+            nextFragment = selectionSortNext(pixelIdx, depth, lastPtr);
+            int nextFragType = getPixelDataType(nextFragment);
+            float nextFragDepth = depth;
+            if (nextFragType == 0) {
+                meshFrag = uncompressMeshPixelData(nextFragment);
+                nextFragDepth = meshFrag.depth;
+            } else if(nextFragType == 1) {
+                volumeFrag = uncompressVolumePixelData(nextFragment);
+                nextFragDepth = volumeFrag.depth;
+            }
+            fragType = nextFragType;
+
+            // transform depth to world space, align sampling positions with camera
+            // NOTE: alignment leads to artifacts during camera transformations due to 
+            // camera-fixated sampling positions. Especially noticeable during coarse sampling.
+            float tStart = ceil(depthToWorld(camera, screenPos, depth) / tIncr) * tIncr;
+            // tStart = depthToWorld(camera, screenPos, depth);
+            float tEnd = depthToWorld(camera, screenPos, nextFragDepth);
+
+            // integrate ray segments of all volumes between current depth and next fragment in world space
+            float t = tStart;
+            while (t < tEnd) {
+                const vec3 samplePosWorld = camera.position + rayDir * t;
+
+                for(int volumeIndex = 0; volumeIndex < 2; ++volumeIndex) {
+                    if(raycastingInfos[volumeIndex].isActive) {
+                        float scalar = 0.0;
+                        // texture sampling cannot be moved into a separate function due to the texture lookup 
+                        // depending on volumeIndex. 
+                        // Arrays of samplers can only be accessed by compile-time integral constant expressions.
+                        // see https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)#Opaque_arrays
+                        switch (volumeIndex) {
+                            case 0:
+                                scalar = sampleVolume(volumeSamplers[0], volumeIndex, samplePosWorld);
+                                break;
+                            case 1:
+                                scalar = sampleVolume(volumeSamplers[1], volumeIndex, samplePosWorld);
+                                break;
+                            case 2:
+                                scalar = sampleVolume(volumeSamplers[2], volumeIndex, samplePosWorld);
+                                break;
+                            case 3:
+                                scalar = sampleVolume(volumeSamplers[3], volumeIndex, samplePosWorld);
+                                break;
+                            default:
+                                break;
+                        }
+
+                        vec4 color = applyTF(tfSamplers[volumeIndex], scalar);
```

Comment:
why is tfSamplers[volumeIndex], any different from volumeSamplers[0]? both should require constants expressions?

Reply:
Good catch. The switch-case was a left-over from passing sampler indices to the function. Now samplers are forwarded.